### PR TITLE
Use endpoint dns if hostname not specified for the address

### PR DESCRIPTION
**Pull request by elenavillamil**
This Pr address incorrect assumption done on [this PR](https://github.com/Azure/Bridge-To-Kubernetes/pull/78/files#diff-c415242ad7d5ebbf107822a6ba3ee5f15ccbbc0220f5948047db4f78bbe1add5). If address did not have hostname it was being skipped, but kubernetes still adds dns for it, it uses the endpoint dns if hostname not specified.

UTs have been updated accordingly.  Fixed also verified against set up that reproduced the issue.

Address #167 and #155.

### CODE REVIEW

File: src/library.tests/WorkloadInformationProviderTests.cs [hunk 1]
```diff
@@ -33,13 +33,13 @@ public WorkloadInformationProviderTests()
...
         [InlineData(10, 3)]
-        [InlineData(1, 3)]
+        [InlineData(2, 3)]
         public async void GetReachableServicesAsync_HeadlessService(int numServices, int numAddresses)
         {
             ConfigureHeadlessService(numServices: numServices, namingFunction: (i) => $"myapp-{i}", numAddresses: numAddresses, addressHostNamingFunction: (i) => $"Host-{i}");
             var result = await _workloadInformationProvider.GetReachableEndpointsAsync(namespaceName: "", localProcessConfig: null, includeSameNamespaceServices: true, cancellationToken: default(CancellationToken));
             // Doing numServices-1 when calculating because we are adding empty subset for one service and that will be skipped
-            Assert.Equal((numServices-1) * (numAddresses), result.Count());
+            Assert.Equal(numAddresses == 0 ? numServices : (numServices-1) * (numAddresses), result.Count());
```
Comment (hsubramanianaks) [inline src/library.tests/WorkloadInformationProviderTests.cs:42]: I don't see use case where numAddress is zero from InlineData, is it made zero somewhere else. ? Please let me know.

File: src/library.tests/WorkloadInformationProviderTests.cs [hunk 2]
```diff
@@ -33,13 +33,13 @@ public WorkloadInformationProviderTests()
         [InlineData(1, 100)]
         [InlineData(5, 20)]
         [InlineData(10, 3)]
-        [InlineData(1, 3)]
+        [InlineData(2, 3)]
         public async void GetReachableServicesAsync_HeadlessService(int numServices, int numAddresses)
         {
             ConfigureHeadlessService(numServices: numServices, namingFunction: (i) => $"myapp-{i}", numAddresses: numAddresses, addressHostNamingFunction: (i) => $"Host-{i}");
             var result = await _workloadInformationProvider.GetReachableEndpointsAsync(namespaceName: "", localProcessConfig: null, includeSameNamespaceServices: true, cancellationToken: default(CancellationToken));
             // Doing numServices-1 when calculating because we are adding empty subset for one service and that will be skipped
```
Comment (hsubramanianaks) [inline src/library.tests/WorkloadInformationProviderTests.cs:41]: Is this comment relevant now, because now dns name can be empty for headless services?

File: src/library/Connect/WorkloadInformationProvider.cs [hunk 3]
```diff
@@ -480,21 +480,33 @@ private int GetServicePortValue(V1ServicePort servicePort, V1Container container
...
-                    foreach (var address in addresses)
+                    
+                    foreach (var address in  subset.Addresses)
                     {
+                        if (address == null) {
+                            continue;
+                        }
+                        string dns = "";
+                        // If hostname is empty for the address, then dns used is that of the endpoint
+                        if (!string.IsNullOrWhiteSpace(address.Hostname))
```
Comment (hsubramanianaks) [inline src/library/Connect/WorkloadInformationProvider.cs:495]: This is suggestion: we can do as a switch statement here instead of if statements as the user suggested ?

File: src/library/Connect/WorkloadInformationProvider.cs [hunk 4]
```diff
@@ -480,21 +480,33 @@ private int GetServicePortValue(V1ServicePort servicePort, V1Container container
...
+                        {
+                            dns = isInWorkloadNamespace ?
+                                    $"{address.Hostname}.{endpoint.Metadata.Name}" :
+                                    $"{address.Hostname}.{endpoint.Metadata.Name}.{endpoint.Metadata.Namespace()}";
+                        }
+                        else
+                        {
+                            dns = isInWorkloadNamespace ?
```
Comment (hsubramanianaks) [inline src/library/Connect/WorkloadInformationProvider.cs:503]: can you add UT for this if/else code block? ex: Assert on the List<EndpointInfo> using CompareLists or for each of this and compare the values ?

**Review by hsubramanianaks — CHANGES_REQUESTED** [review]
Left some comments and do you think we should add a separate task/story to create todo-app with headless service so that this can be tested as well by testing team members ? Since we are seeing bugs around this often. Please let me know your thoughts. Thanks.

File: src/library.tests/WorkloadInformationProviderTests.cs (hunk 1 above)
Comment (elenavillamil) [reply src/library.tests/WorkloadInformationProviderTests.cs:42]: Ah this was left over from when I was working on the UT. Will remove.

File: src/library/Connect/WorkloadInformationProvider.cs (hunk 3 above)
Comment (elenavillamil) [reply src/library/Connect/WorkloadInformationProvider.cs:495]: I decided agains switch because number of conditions is not large and readability seemed better this way. https://www.geeksforgeeks.org/switch-vs-else/ this is some reading I did on switch vs if. It seems switch is preffer if number if else nested cases is large.

File: src/library/Connect/WorkloadInformationProvider.cs (hunk 4 above)
Comment (elenavillamil) [reply src/library/Connect/WorkloadInformationProvider.cs:503]: done

**Comment by elenavillamil** [comment]
> Left some comments and do you think we should add a separate task/story to create todo-app with headless service so that this can be tested as well by testing team members ? Since we are seeing bugs around this often. Please let me know your thoughts. Thanks.

Yes, I think the task should be added under E2E testing story. Once we have automated testing it will be a lot easier to add more coverage of this type of edge cases. But I dont really want to add more load on testing team atm.

File: src/library.tests/WorkloadInformationProviderTests.cs [hunk 5]
```diff
@@ -30,19 +30,29 @@ public WorkloadInformationProviderTests()
...
-        public async void GetReachableServicesAsync_HeadlessService(int numServices, int numAddresses)
+        [InlineData(1, 100, true)]
+        [InlineData(5, 20, false)]
+        [InlineData(10, 3, true)]
+        [InlineData(2, 3, false)]
+        public async void GetReachableServicesAsync_HeadlessService(int numServices, int numAddresses, bool isInWorkloadNamespace)
         {
-            ConfigureHeadlessService(numServices: numServices, namingFunction: (i) => $"myapp-{i}", numAddresses: numAddresses, addressHostNamingFunction: (i) => $"Host-{i}");
-            var result = await _workloadInformationProvider.GetReachableEndpointsAsync(namespaceName: "", localProcessConfig: null, includeSameNamespaceServices: true, cancellationToken: default(CancellationToken));
+            var expected = ConfigureHeadlessService(numServices: numServices, namingFunction: (i) => $"myapp-{i}", numAddresses: numAddresses, addressHostNamingFunction: (i) => $"Host-{i}", isInWorkloadNamespace);
```
Comment (hsubramanianaks) [inline src/library.tests/WorkloadInformationProviderTests.cs:39]: can we change the variables expected to expectedServicesList and instead of result, finalServicesList since many of the variables are defined ```var``` it is confusing when someone reads the code. Just a nice to have suggestion. Other than than looks good, I like the idea of covering all different use cases programatically in the test using InlineData :). 

or another suggestion instead of var result, you can change it to ```IEnumerable<EndpointInfo> result ``` , same for expected variable. 

Thanks.

**Review by hsubramanianaks — APPROVED** [review]
LGTM except a minor comment,

File: src/library.tests/WorkloadInformationProviderTests.cs (hunk 5 above)
Comment (elenavillamil) [reply src/library.tests/WorkloadInformationProviderTests.cs]: sure, done.

**Review by hsubramanianaks — APPROVED** [review]
LGTM
